Feat/backend v1

suryanshkushwaha opened the pull request:
This pull request introduces the implementation of CRUD (Create, Read, Update, Delete) operations along with a RESTful API for our application. The API is designed to facilitate efficient data management and interaction with the underlying database.

**Key Features:**
- **Create:** Added functionality to create new records in the database.
- **Read:** Implemented endpoints to retrieve data, including support for filtering and pagination.
- **Update:** Enabled the ability to modify existing records through specific endpoints.
- **Delete:** Added the capability to delete records from the database.
- **REST API:** Structured the API following REST principles, ensuring proper use of HTTP methods (GET, POST, PUT, DELETE) and status codes.

suryanshkushwaha committed:
Update source.sh

Changed the literals to strings

Mahidharah committed:
Setup basic mongoDB connection in the backend

- Add packages
- Add functional code in index.ts file for mongoDB
- add db.ts file in config

suryanshkushwaha committed:
add: template directories

Created separate directories with templates for REST API implementation

suryanshkushwaha committed:
feat(api): Add REST API Structure

Implemented REST API structure, including:
Models: defined data schema
Controllers: implemented logic
Routes: configured API endpoints
Server (server.ts): setup API server

suryanshkushwaha committed:
add: basic test cases

bhnuka commented on backend/src/__test__/controllers/questionController.test.ts:
```diff
@@ -0,0 +1,117 @@
+  // Test: Fetch a question by ID
+  it('should fetch a question by ID', async () => {
+    const question = new Question({
+      questionId: 1,
```
we should standardise if we're using questionId or Id. For frontend we used Id.

suryanshkushwaha replied:
Noted. Resolved!

bhnuka commented on backend/src/__test__/controllers/questionController.test.ts:
```diff
@@ -0,0 +1,117 @@
+      questionId: 1,
+      title: 'What is JavaScript?',
+      description: 'A programming language.',
+      category: 'Programming',
```
should be categories, not category (example questions given in document have multiple categories)

bhnuka commented on backend/src/models/questionModel.ts:
```diff
@@ -0,0 +1,24 @@
+    questionId: number;
+    title: string;
+    description: string;
+    category: string;
```
since this is categories, should be string[]

Yongbeom-Kim replied:
if we want categories, we should actually have a collection of available categories, and then check whether the category belongs to a document in the collection, and throw if it does not belong. Thoughts @suryanshkushwaha @Mahidharah  ? Can think about it after M2

Mahidharah replied:
I agree, since we have a fixed set of categories I assume, and this should be part of business logic. 

However, do you intend of allowing authorised users to add categories? Then this might change things, unless we allow them to add to the category collection.

bhnuka reviewed (changes requested):
just need to come to both agree on using questionId or Id and we should use categories rather than category (plural). handling of category also changes slightly to handle multiple categories.

bhnuka commented on backend/yarn.lock:
by right i think we should stick to either yarn.lock or package-lock for each subdirectory (i.e. in frontend we only have yarn.lock) having both may make things confusing at times. anyone else, feel free to correct me if im wrong

Yongbeom-Kim replied:
Yes, ideally should stick to yarn.lock, and if we have a yarn.lock, we cannot have a package-lock

Yongbeom-Kim commented on backend/src/server.ts:
```diff
@@ -0,0 +1,34 @@
+
+// Middleware for handling CORS and JSON parsing
+app.use(cors({
+  origin: 'http://localhost:3000',
```
Vite default port is 5173? If we use 3000 it will fail on vite, right? Does ts cors accept wildcard? Propose to change to localhost:* if possible to avoid unnecessary pain, or add localhost:5173 at least

Mahidharah replied:
Suryansh and I used postman for testing, which explains the port number. Will try wildcards, else will include both for testing and for integrating with frontend

Yongbeom-Kim commented on source.sh:
```diff
@@ -4,9 +4,9 @@ PWD="$(git rev-parse --show-toplevel)"
+source "$PWD/.env"
 
-GOOGLE_APPLICATION_CREDENTIALS=$(realpath $PWD/$GOOGLE_APPLICATION_CREDENTIALS)
+GOOGLE_APPLICATION_CREDENTIALS=$(realpath "$PWD/$GOOGLE_APPLICATION_CREDENTIALS")
```
```suggestion
GOOGLE_APPLICATION_CREDENTIALS=$(realpath "$PWD/$GOOGLE_APPLICATION_CREDENTIALS")
```
```suggestion
GOOGLE_APPLICATION_CREDENTIALS="$(realpath "$PWD/$GOOGLE_APPLICATION_CREDENTIALS")"
```

Mahidharah committed:
Add sample data

- Added sample data and a script to add sample data, if DB doesnt exist, or if DB exists just skip the adding of sample data

- Commented some sample data for testing of API calls

- Fix category to categories

However, the change to categories is not passing a post request to add a sample question via api call, hence will need some debugging to ensure string is converted to array correctly

Mahidharah committed:
Fix array handling for multiple categories

- Now post requests work, where in my previous commit it was not working
- Does some checking before calling split for categories

suryanshkushwaha committed:
feat: add autoincrement questionId

Implemented counter that automatically increments on each new question entry

suryanshkushwaha committed:
Merge branch 'feat/backend-v1' of https://github.com/CS3219-AY2425S1/cs3219-ay2425s1-project-g10 into feat/backend-v1

Mahidharah committed:
Update questionController.ts

- Bug fixes

suryanshkushwaha committed:
minor fix

Renamed 'complexity' to 'difficulty'

suryanshkushwaha committed:
fix: sample data to use server for creation

suryanshkushwaha committed:
minor fixes

Minor fix and removed test cases

Mahidharah committed:
Set up MongoDB on cloud

- Set up MongoDB in cloud
- Set read/write permissions to all

Mahidharah committed:
Add uri to test

Mahidharah committed:
Delete package-lock.json

Mahidharah committed:
Merge branch 'dev' into feat/backend-v1

Mahidharah reviewed (approved):
LGTM

bhnuka commented:
LGTM!

bhnuka reviewed (approved):
LGTM!
